lib/key/check.js

```diff
@@ -1,6 +1,102 @@
+ * @param {Set<openpgp.enums.publicKey} requirements.allowedPublicKeyAlgorithms
+ * @param {Set<openpgp.enums.curve} requirements.allowedCurves
+ * @throws {Error} if the kay is considered too weak
+ */
+function checkKeyStrength(publicKeyOrSubkey, requirements) {
+    const { allowedCurves, allowedPublicKeyAlgorithms, minRSABits } = requirements;
+    const keyInfo = publicKeyOrSubkey.keyPacket.getAlgorithmInfo();
+    const keyAlgo = openpgp.enums.write(openpgp.enums.publicKey, keyInfo.algorithm);
+
+    if (!allowedPublicKeyAlgorithms.has(keyAlgo)) {
+        throw new Error(`${keyInfo.algorithm} keys are considered unsafe`);
+    }
+
+    if (keyInfo.curve && !allowedCurves.has(keyInfo.curve)) {
+        throw new Error(`Keys using curve ${keyInfo.curve} are considered unsafe`);
+    }
+
+    if (keyInfo.rsaBits && keyInfo.rsaBits < minRSABits) {
+        throw new Error(`Keys shorter than ${minRSABits} bits are considered unsafe`);
+    }
+}
+
+/**
+ * Checks whether the primary key and the encryption subkey are strong enough to be imported as personal keys.
+ * These checks are stronger than those for contact keys.
+ * @param {OpenPGPKey} publicKey - key to check
+ * @throws {Error} if the key is considered too weak
+ * @async
+ */
+export async function checkPersonalKeyStrength(publicKey) {
+    const { enums } = openpgp;
+    // primary key
+    checkKeyStrength(publicKey, {
+        allowedPublicKeyAlgorithms: new Set([
+            enums.publicKey.rsa_encrypt_sign,
+            enums.publicKey.rsa_sign,
+            enums.publicKey.ecdsa,
+            enums.publicKey.eddsa
+        ]),
+        allowedCurves: new Set([enums.curve.ed25519, enums.curve.p256, enums.curve.p384, enums.curve.p521]),
+        minRSABits: 2048
+    });
+    // TODO should we check all subkeys instead? (faster, but we dont know if they are signing or encryption subkeys)
+    const encryptionSubkey = await publicKey.getEncryptionKey(); // for Proton keys, the primary key is always a sign-only key
+    checkKeyStrength(encryptionSubkey, {
+        allowedPublicKeyAlgorithms: new Set([
+            enums.publicKey.rsa_encrypt_sign,
+            enums.publicKey.rsa_encrypt,
+            enums.publicKey.ecdh
+        ]),
+        allowedCurves: new Set([enums.curve.curve25519, enums.curve.p256, enums.curve.p384, enums.curve.p521]),
+        minRSABits: 2048
+    });
+}
+
+/**
+ * Checks whether a primary key is strong enough to be used for message encryption.
+ * @param {OpenPGPKey} publicKey - key to check
+ * @throws {Error} if the key is considered too weak
+ * @async
+ */
+export async function checkContactKeyStrength(publicKey) {
+    const { enums } = openpgp;
+    const minRSABits = 1024;
+    const allowedCurves = new Set([
+        enums.curve.ed25519,
+        enums.curve.curve25519,
+        enums.curve.p256,
+        enums.curve.p384,
+        enums.curve.p521
+    ]);
+
+    // primary key only (TODO does this matter for contacts?)
```

Comment:
?

Reply:
Yes, for verifying signatures. IMHO, can't we just do the same thing here? IMHO, we should also warn for 1024-bit contact keys.

Reply:
What same thing?
Contact key warnings are going to be shown in the composer, so I think we already agreed we should leave 2048 for now.

Reply:
OK, fair enough. I mean, I think eventually we should, but yeah we might need a transitional period where we don't since we generated those ourselves in the past. In any case, then I think the two functions can be essentially the same except for the `minRsaBits`, right?

Reply:
If we don't require encryption key presence, then yes, but since the use-case is different, I'd still expose two different functions? (so that the caller does not need to set params)